va/validation-authority.go

```diff
@@ -717,50 +722,67 @@ func newCAASet(CAAs []*dns.CAA) *CAASet {
 	return &filtered
 }
 
-func (va *ValidationAuthorityImpl) getCAASet(ctx context.Context, hostname string) (*CAASet, error) {
-	hostname = strings.TrimRight(hostname, ".")
-	labels := strings.Split(hostname, ".")
-
-	// See RFC 6844 "Certification Authority Processing" for pseudocode.
-	// Essentially: check CAA records for the FDQN to be issued, and all
-	// parent domains.
-	//
-	// The lookups are performed in parallel in order to avoid timing out
-	// the RPC call.
-	//
-	// We depend on our resolver to snap CNAME and DNAME records.
+type caaResult struct {
+	records []*dns.CAA
+	err     error
+}
 
-	type result struct {
-		records []*dns.CAA
-		err     error
+func parseResults(results []caaResult) (*CAASet, error) {
+	// Return first result
+	for _, res := range results {
+		if res.err != nil {
+			return nil, res.err
+		}
+		if len(res.records) > 0 {
+			return newCAASet(res.records), nil
+		}
 	}
-	results := make([]result, len(labels))
+	return nil, nil
+}
 
+func (va *ValidationAuthorityImpl) parallelCAALookup(ctx context.Context, name string, lookuper func(context.Context, string) ([]*dns.CAA, error)) []caaResult {
+	labels := strings.Split(name, ".")
+	results := make([]caaResult, len(labels))
 	var wg sync.WaitGroup
 
 	for i := 0; i < len(labels); i++ {
 		// Start the concurrent DNS lookup.
 		wg.Add(1)
-		go func(name string, r *result) {
-			r.records, r.err = va.DNSResolver.LookupCAA(ctx, name)
+		go func(name string, r *caaResult) {
+			r.records, r.err = lookuper(ctx, name)
 			wg.Done()
 		}(strings.Join(labels[i:], "."), &results[i])
 	}
 
 	wg.Wait()
+	return results
+}
 
-	// Return the first result
-	for _, res := range results {
-		if res.err != nil {
-			return nil, res.err
-		}
-		if len(res.records) > 0 {
-			return newCAASet(res.records), nil
-		}
-	}
+func (va *ValidationAuthorityImpl) getCAASet(ctx context.Context, hostname string) (*CAASet, error) {
+	hostname = strings.TrimRight(hostname, ".")
 
-	// no CAA records found
-	return nil, nil
+	// See RFC 6844 "Certification Authority Processing" for pseudocode.
+	// Essentially: check CAA records for the FDQN to be issued, and all
+	// parent domains.
+	//
+	// The lookups are performed in parallel in order to avoid timing out
+	// the RPC call.
+	//
+	// We depend on our resolver to snap CNAME and DNAME records.
+
+	results := va.parallelCAALookup(ctx, hostname, va.DNSResolver.LookupCAA)
+	set, err := parseResults(results)
+	if err == nil {
+		return set, nil
+	}
+	if va.caaPR == nil {
+		return nil, err
+	}
+	// we have a caaPublicResolver and one of the local lookups failed
+	// so we talk to the Google Public DNS service over various VPN
+	// interfaces instead
+	results = va.parallelCAALookup(ctx, hostname, va.caaPR.LookupCAA)
```

Comment:
Before calling out to the public resolver, check that the error was a timeout. Other errors get return immediately.